### PR TITLE
Bugfix: lax_deserialize currently accepts deserializing from NaN

### PR DESCRIPTION
**Pull request by FSMaxB**
With the `lax_deserialize` feature, `NaN` currently deserializes to `0`. I don't think this was intended.

Although this can't happen with `serde_json` because JSON doesn't support `NaN` etc., this can very well happen with other deserializers.

This pull request fixes this by declinining to deserialize `NaN`, `Infinity` and `-Infinity`.

Note that using `IntoDeserializer`, as I've done here, could be used for the other tests as well, thereby getting rid of the necessity to use `serde_json` if desired. (This, by extension, should make the tests work with `no-std` again, thereby making the workaround for https://github.com/serde-rs/json/pull/516 obsolete as well ...).

### CODE REVIEW

**Comment by jplatte** [comment]
> This pull request fixes this by declinining to deserialize `NaN`, `Infinity` and `-Infinity`.

To be clear, the infinities were already rejected before, right?



> Note that using `IntoDeserializer`, as I've done here, could be used for the other tests as well, thereby getting rid of the necessity to use `serde_json` if desired. (This, by extension, should make the tests work with `no-std` again, thereby making the workaround for [serde-rs/json#516](https://github.com/serde-rs/json/pull/516) obsolete as well ...).

Sounds good!

**Comment by FSMaxB** [comment]
> To be clear, the infinities were already rejected before, right?

You're right, I've missed that. I've changed the code to only check for `is_nan`.

> Sounds good!

I'll do that then.

**Comment by FSMaxB** [comment]
Ok, this was a bit more complicated than anticipated because I totally forgot about the `Serialize` direction. But nothing a custom `TestSerializer` couldn't fix!

**Comment by jplatte** [comment]
There's also `serde_test` but I find its API to be a little annoying to use.

**Comment by jplatte** [comment]
Maybe it would be good for this use case though.

**Comment by FSMaxB** [comment]
Sorry about the force push, I still had `expand.rs` in there (I was looking at what `#[derive(Serialize)]` produces.

> There's also serde_test but I find its API to be a little annoying to use.
> Maybe it would be good for this use case though.

I'll look into that. Let's hope that it works with `no-std` :grimacing:

File: tests/int.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use serde_test::{assert_ser_tokens, Token};
+
+#[test]
+fn serialize_int() {
+    assert_serialize(100);
+    assert_serialize(0);
+    assert_serialize(-100);
+}
+
+fn assert_serialize(number: i32) {
```
Comment (FSMaxB) [inline tests/int.rs:14]: I was tempted to use `#[track_caller]` here but then I remembered that the MSRV is 1.35.

Using a macro would also be possible but I don't think it's really worth it.

File: tests/int.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    assert_ser_tokens(
+        &Int::from(number),
+        &[Token::NewtypeStruct { name: "Int" }, Token::I64(number as _)],
```
[the remark this answers is not in the document]
Comment (jplatte) [reply tests/int.rs:17]: That's a good point. Idk which serializers out there actually serialize sth. for `NewtypeStruct`, but I agree that we shouldn't emit it. If the derives aren't to be replaced entirely, `#[serde(transparent)]` could be used to get rid of it.

**Comment by jplatte** [comment]
I have merged the first two commits (as one squashed commit) for now. It looks like you will have to rebase, which is a bit weird to me since I would expect a rebase to go through w/o any manual work needed.

**Comment by FSMaxB** [comment]
I'll close this pull request then and create a new one for the testing changes. I expect that this requires an interactive rebase, but no worries, I have experience with that :grin: .

**Pull request closed without merging**